modules/primo/primo.module

```diff
@@ -0,0 +1,201 @@
+    '#type' => 'textfield',
+    '#title' => t('Institution code'),
+    '#description' => t('Relevant for restricted scopes or for when searching against Primo Central.'),
+    '#required' => TRUE,
+    '#default_value' => variable_get('primo_institution_code', ''),
+  );
+
+  $form['primo']['primo_enable_logging'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Enable logging'),
+    '#default_value' => variable_get('primo_enable_logging', FALSE),
+    '#description' => t('Logs requests to the Primo webservice.'),
+  );
+
+  $form['covers'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Covers'),
+    '#tree' => FALSE,
+  );
+
+  $form['covers']['primo_covers_hostname'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Hostname'),
+    '#description' => t('The hostname for which to include covers.'),
+    '#required' => TRUE,
+    '#default_value' => variable_get('primo_covers_hostname',
+      PRIMO_COVERS_DEFAULT_HOSTNAME
+    ),
+  );
+
+  return system_settings_form($form);
+}
+
+/**
+ * Implementation of hook_ting_covers().
+ */
+function primo_ting_covers($entities) {
+  $ids = array_map(function(TingEntity $entity) {
+    return $entity->ding_entity_id;
+  }, $entities);
+
+  // Replace entity ids with a set of Primo ids which are known to have covers.
+  // TODO: Remove this once we have proper search integration with Primo.
+  $ids = array(
+    'ICE01_PRIMO_TEST000811009',
+    'ICE01_PRIMO_TEST001027834',
+    'ICE01_PRIMO_TEST000335899',
+  );
+
+  $document = primo_load_by_recordid($ids);
+
+  $thumbnail_urls = array_fill_keys($ids, NULL);
+  array_walk($thumbnail_urls, function(&$url, $id, DOMDocument $document) {
+    $urls = primo_extract_thumbnail_urls($document, $id);
+    // We can only handle a single thumbnail per id so return the first.
+    $url = array_pop($urls);
+  }, $document);
+
+  // Splice the results back into an array mapping original ids with thumbnail
+  // urls for Primo ids.
+  // TODO: Remove this once we have proper search integration with Primo.
+  $thumbnail_urls = array_slice($thumbnail_urls, 0, count($entities));
+  return array_combine(
+    array_keys($entities),
+    array_pad($thumbnail_urls, count($entities), reset($thumbnail_urls))
+  );
+}
+
+/**
+ * Load documents based on their record id.
+ *
+ * @param string[] $rids
+ *   An array of Primo record ids.
+ *
+ * @return \DOMDocument
+ *   A DOMDocument containing the resulting XML. This XML will contain elements
+ *   for each of the requested record ids.
+ */
+function primo_load_by_recordid(array $rids) {
+  $url = variable_get('primo_base_url') . '/PrimoWebServices/xservice/search/brief';
+  $params = array(
+    'institution' => variable_get('primo_institution_code'),
+    'query' => 'rid,contains,' . implode($rids, ' OR '),
+    'indx' => 1,
+    'bulkSize' => count($rids),
+  );
+
+  $result = drupal_http_request(url($url, array('query' => $params)));
+
+  $document = new DOMDocument();
```

Comment:
If an error occured we should probably just log it and don't try to parse the document. As far as I remember $result->error can be used to check the status.

Reply:
Agreed. Some error handling is needed here.